File: C++/382.cpp
__________________________________________________________________________________________________
sample 28 ms submission
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
const static int _= []()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    Solution(ListNode* head) {
        srand (time(NULL));
        m_head = head;
    }
    
    /** Returns a random node's value. */
    int getRandom() {
        int guess = rand();
        ListNode * runner = m_head;
        int listSize=1;
        bool largerThanList = false;
        for (int i=0; i<guess; i++)
        {            
            if (runner->next)
                runner=runner->next;
            else
            {
                largerThanList=true;
                break;
            }                
            listSize++;
        }
        if (largerThanList)
        {
            int n = guess%listSize;
            runner = m_head;
            for (int i=0; i<n; i++)
                runner = runner->next;
        }
        
        return runner->val;
    }
private:
    ListNode * m_head;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */
__________________________________________________________________________________________________
sample 16160 kb submission
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    Solution(ListNode* head) {
        len=0;
        ListNode *cur=head;
        this->head=head;
        while(cur)
        {
            ++len;
            cur=cur->next;
        }
    }
    
    /** Returns a random node's value. */
    int getRandom() {
        int t=rand()%len;
        ListNode*cur=head;
        while(cur&&t)
        {
            t--;
            cur=cur->next;
        }
        return cur->val;
    }
private:
    int len;
    ListNode* head;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */
__________________________________________________________________________________________________

File: Python3/382.py
__________________________________________________________________________________________________
sample 72 ms submission
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
import random


class Solution:

    def __init__(self, head: ListNode):
        self.head = head
        self.count = self.to_count()
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        """
        

    def getRandom(self) -> int:
        node = self.head
        p = random.random()//(1/self.count)
        for i in range(int(p)):
            node = node.next
        return node.val
        """
        Returns a random node's value.
        """
    
    def to_count(self):
        node = self.head
        count = 0
        while node is not None:
            count += 1
            node = node.next
        return count
    
# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
__________________________________________________________________________________________________
sample 15976 kb submission
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

from random import random

class Solution:

    def __init__(self, head: ListNode):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        """
        self.head = head
        pass

    def getRandom(self) -> int:
        """
        Returns a random node's value.
        """
        max_p = -1
        cur = self.head
        res = 0
        while cur:
            cur_p = random()
            if cur_p > max_p:
                res = cur.val
                max_p = cur_p
            cur = cur.next
                
        return res
        
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
__________________________________________________________________________________________________
